packages/runner/src/context.ts

```diff
@@ -102,14 +106,34 @@ export function withTimeout<T extends (...args: any[]) => any>(
   }) as T
 }
 
+const abortControllers = new WeakMap<TestContext, AbortController>()
+
+export function abortIfTimeout([context]: [TestContext?], error: Error): void {
+  if (context) {
+    abortContextSignal(context, error)
+  }
+}
+
+export function abortContextSignal(context: TestContext, error: Error): void {
+  const abortController = abortControllers.get(context)
+  abortController?.abort(error)
+}
+
 export function createTestContext(
   test: Test,
   runner: VitestRunner,
 ): TestContext {
   const context = function () {
     throw new Error('done() callback is deprecated, use promise instead')
-  } as unknown as TestContext
+  } as unknown as WriteableTestContext
+
+  const abortController = abortControllers.get(context) || (() => {
+    const abortController = new AbortController()
+    abortControllers.set(context, abortController)
+    return abortController
+  })()
```

Comment:
Nitpick: Personally I would like this IIFE to be just good old `if`. It's okey to be verbose to reduce complexity 😄 

```suggestion
  let abortController = abortControllers.get(context)
  
   if(!abortController) {
    abortController = new AbortController()
    abortControllers.set(context, abortController)
  }
```